C++ front end fixes [blocks: #2554]

tautschnig opened the pull request:
This is a series of fixes mainly compiled by @peterschrammel plus some additional work of mine.

This includes the following commits to be reviewed:
- [x] ~e92421da24 Do not use assert without prior declaration~
- [ ] ~be13461761 Fix syntax errors in C++ regression tests~
- [x] ~c9094dd044 Fix broken comment frame~
- [x] ~12a71d181c Do not check friend declarations in wrong scope (moved to #2318)~
- [x] ~6ffc37c27d Do not add method body on declaration (will be handled on use)~
- [x] ~4576f14629 Line break~
- [ ] ~6dc9b9babc Fix delayed method body conversion for templates~
- [x] ~1db4a0459c Initial set of mini-system-c tests~
- [ ] ~c9f33342b8 Clean up unused template instantiation symbols~
- [ ] ~dc7b8c882e Fix disambiguation~
- [x] ~ba0df47444 C++ front-end: store typedef names (moved to #4551)~
- [x] ~43c0aa62f3 C++ front-end: parse GCC attributes~
- [ ] ~cafaac5947 Permit asm post-declarator mixed in any order with other qualifiers~
- [ ] ~6c6618b1f6 Permit GCC attributes after struct/class in class declaration~
- [ ] ~da7822bf7b Mark already-typechecked types as such~
- [x] ~f9e723e3c2 Typecheck initializer lists~
- [ ] ~d1bf14aad2 C++ front-end: Use C factory for compiler builtins~
- [x] ~5c14d094f8 Do not lose method qualifiers~
- [x] ~a67c62b99d Whitespace cleanup, comment type fixed~
- [x] ~f165ca9b9e Fix operator parsing~
- [x] ~d351125d4c Failing regression test from #661~
- [x] ~d3c2f9d2a8 Failing regression test from #933~
- [x] ~9fdefe8593 Enable compilation with DEBUG defined. Some of the code inside #ifdef DEBUG ... #ENDIF had become stale (moved to #4475)~
- [ ] ~edc0cd8774 Swap order of subtypes in construction of merged_type nodes, so that the main type is at the end. Handle and record GCC function attribute noreturn.~
- [x] ~f89a086304 Style improvements (moved to #4475)~
- [x] ~7cac0cecfc Prevent regression failures of cpp tests on windows.~
- [x] ~81a5569c73 Do not lint .h files in regression/~
- [x] ~a6e83f6b8f Don't abbreviate variable names (moved to #4552)~
- [x] 7aaaa8258d Mark parameters (and their symbols) as such (seems unnecessary)
- [ ] 63f0ef87c1 Do not mark arbitrary constants as macros
- [ ] d425cd7613 Declarator -> symbol conversion to match C implementation
- [ ] 5467da39fa C++: Remove internal symbols (macros in particular) before linking
- [x] ~44cb7c9892 decltype(bit field type) is the underlying subtype~
- [ ] ~c0a4278b84 Move is_template_scope to parent as it holds the id_class member~
- [ ] ~ab0812c47f Factor out set-up of sub_scope for template instantiation~
- [ ] ~87fddb8c3d Properly set up scope prefix instead of hacking it away~
- [x] ~0c686a8ffb Remove duplicate save_scope~
- [x] ~b3329e83f2 Remove redundant invariants~

tautschnig commented:
I will need to make the linter happy and figure out how to support the Visual Studio-provided syntax.

andreast271 commented:
I am currently looking at the failing checks. I fixed the linter complaints outside of regression, but there will be new ones, because assert() must be replaced by INVARIANT or one of its derivatives. 
There are also 9 failing regression tests, all in the cpp directory, and all with the same error message. I plan to look at that next week, using branch peterschrammel:systemc as baseline, since at least one of the test is not failing on peterschrammel:systemc.

andreast271 commented:
I have a fix for the compilation errors with that happen with -DDEBUG.

martin-cs commented:
@andreast271 ( and @tautschnig and @peterschrammel ) thank you for looking at this; this is incredibly valuable work.

andreast271 commented on src/cpp/parse.cpp:
```diff
@@ -2102,28 +2104,204 @@ bool Parser::optAlignas(typet &cv)
+      merge_types(attr, t);
+      break;
+    }
+
```
This creates a merged type with op0 = function_type, op1 = noreturn. 
Is that intended? Typecheck code assumes that following subtypes uses the back of the list for merged types. Here this would be the noreturn node, which seems odd.
Adding the noreturn attribute in this way will currently cause an assertion failure in cpp_declaratort::merge_type for regression test cases that #include assert.h. 
Even after a fix in cpp_declaratort::merge_type the type would then fail to typecheck in cpp_convert_plain_type(typet &type)
I'm happy to work on fix after getting an explanation of the intended use of merged_type in C++ frontend.

tautschnig replied:
I believe I had tried to mimic what the C front-end does. The obvious fix would be to swap the parameters here, but I'm not sure this will result in the attribute being picked up at all?

andreast271 replied:
Where should the attribute end up after type checking? At some point function_type is converted to code_type. At that stage the type qualifiers need to be put somewhere in the code_type node. I didn't see anything in the C++ front end that does that, so right now cpp_convert_plain_type fails because it cannot deal with a qualifier of a function type. I could fix it if I understood how the result of type checking the function should look like.

andreast271 replied:
The merge_types function in the C and C++ front end use a different convention for the order of arguments: in the C front end, it is merge_types(dest, src), while in C++ it is merge_types(src, dest). A possible fix would then be to change the merge_types function in C++ so that it adds the new type at the front of the subtypes (or rather anywhere but at the end). Thoughts?

andreast271 replied:
Changing the merge_types functions seems to work. Regression passes and I also have a fix for the -DDEBUG compilation failures and some of the linter issues. What is the process for gettings these fixes into the PR?

tautschnig replied:
Could you check whether you are able to push to my branch? If you aren't, you can certainly create a pull request in my fork (against that branch), and as soon as I merge it the changes will show up here.

peterschrammel commented:
This would need rebasing...

tautschnig commented:
Rebase done.

andreast271 commented:
The linter failures are all from cpp files in the regression. I discussed this with @thk123 a while ago. He agreed that the linter shouldn't run on regression test sources, and also clarified that linter failures do not prevent a merge. IMO the correct fix is to change the linter so that it skips the regression directory, but that should be a separate PR.

andreast271 commented:
Prefix: I have given up on supporting VS headers in cbmc. I _compile_ cbmc on Windows with VisualStudio (not just the compiler, but the IDE), but even then I _run_ under cygwin using `--gcc --cpp98`, so that preprocessing is done by gcc using `-std=c++98`. Everything else fails as soon as I include a system header file. 

There are 5 test failures in the cpp group on windows reported by the appveyor build:

1. `Function_Overloading2`. This is a KNOWNBUG in develop, so it never ran on windows, so it's not a regression. Solution: remove test from appveyor build
2. `Function_Overloading3`. This is a new test, so it never ran on windows, so it's not a regression. Solution: remove test from appveyor build.
3. to 5. `Resolver10, Resolver11, Template_Parameters1`: These tests were changed to `#include <cassert>` in `main.cpp`. This causes failures on windows - see Prefix above re. VS headers. Solution: since the test sources don't actually contain any assertions, revert the changes. 

@tautschnig : I can create a PR to tautschnig/c++-front-end-fixes with the suggested solutions if you agree with them.

smowton commented on src/cpp/cpp_typecheck_declaration.cpp:
```diff
@@ -6,6 +6,7 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
 \********************************************************************/
 
+
```
Undo

smowton commented on src/cpp/cpp_typecheck_method_bodies.cpp:
```diff
@@ -39,3 +53,26 @@ void cpp_typecheckt::typecheck_method_bodies(
+
+  // We have to prevent the same method to be added multiple times
+  //   otherwise we get duplicated symbol prefixes
+  if(methods_seen.find(_method_symbol->name) != methods_seen.end())
```
Use `insert` and check its return value rather than find-then-insert double-lookup

smowton commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -474,51 +483,44 @@ void cpp_typecheck_resolvet::disambiguate_functions(
+    for(resolve_identifierst::const_iterator
+        it1=old_identifiers.begin();
+        it1!=old_identifiers.end();
         it1++)
```
Might as well switch to pre-increment while we're changing things here

smowton commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -474,51 +483,44 @@ void cpp_typecheck_resolvet::disambiguate_functions(
+        it1!=old_identifiers.end();
         it1++)
     {
+#if 0
```
ifdef DEBUG?

smowton commented on src/cpp/parse.cpp:
```diff
@@ -473,12 +473,19 @@ void Parser::merge_types(const typet &src, typet &dest)
+    // since several analysis functions traverse via the end for
+    // merged_types
+    typet::subtypest &sub=dest.subtypes();
+    sub.emplace(sub.begin(), src);
```
Might as well use insert, since emplace is just calling a copy-constructor itself here plus doing some mildly-costly (in my experience) gymnastics to avoid copying the argument in the meantime

smowton commented on src/goto-programs/goto_convert_functions.cpp:
```diff
@@ -1,4 +1,4 @@
-/**** ***************************************************************\
+/********************************************************************\
```
Accidental change?

andreast271 replied:
No: 03c9894 ("Fix broken comment frame", 2017-07-31)

thk123 commented on regression/cpp/Apple_extensions1/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
If there isn't already, could you raise an issue and add it as a comment at the bottom of test.desc ( @tautschnig )

tautschnig replied:
Done.

thk123 commented on regression/cpp/Template_Specialisation2/main.ii:
```diff
@@ -15,3 +15,6 @@ struct __member_pointer_traits_imp<_Rp (_Class::*)() const>
 {
 };
 
+int main(int argc, char* argv[])
```
Should this become `main.cpp` as with the other file ( @tautschnig )

tautschnig replied:
I don't know -- there isn't really enough information in the initial commit message for that file (e155e42030921), so I'd suggest to leave it as is. It should not have any impact. Should.

thk123 commented on src/cpp/cpp_constructor.cpp:
```diff
@@ -47,12 +47,16 @@ codet cpp_typecheckt::cpp_constructor(
     // The purpose of the tag #array_ini is to rule out ill-formed
     // programs.
 
+    // TODO: in a program it is not allowed, but the implementation of the
```
I don't think this comment is very clear ( @peterschrammel )

peterschrammel replied:
De/activating this code has no effect on cpp and systemc regression tests. We should remove this change from this PR.

tautschnig replied:
I have removed the commit "Deactivate internal array initializer".

thk123 commented on src/cpp/cpp_declarator_converter.cpp:
```diff
@@ -11,6 +11,12 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
 #include "cpp_declarator_converter.h"
 
+//#define DEBUG
```
Nit: don't check this line in ( @peterschrammel )

thk123 commented on src/cpp/cpp_declarator_converter.cpp:
```diff
@@ -11,6 +11,12 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
+//#define DEBUG
+
+#ifdef DEBUG
```
Possibly this is just ending up in the wrong commit - but we don't seem to use `iostream` in this cpp. ( @peterschrammel )

thk123 commented on src/cpp/cpp_declarator_converter.cpp:
```diff
@@ -69,8 +75,18 @@ symbolt &cpp_declarator_convertert::convert(
 
     // check the declarator-part of the type, in that scope
-    cpp_typecheck.typecheck_type(final_type);
+    //TODO: if it is a friend declaration we have to type-check it
```
This comment doesn't seem to match the code? It seems like `if(!is_friend)` then we do the type-check in this scope(?) otherwise we defer to outside `cpp_save_scopet` ( @peterschrammel )

thk123 commented on src/cpp/cpp_scope.cpp:
```diff
@@ -143,7 +143,8 @@ void cpp_scopet::lookup(
   if(!id_set.empty() &&
-     id_class!=id_classt::TEMPLATE) return; // done, upwards scopes are hidden
+     id_class!=id_classt::TEMPLATE)
+    return; // done, upwards scopes are hidden
```
Nit: since the if statement is multiple lines, wrap return in braces ( @peterschrammel  )

andreast271 replied:
Moved if condition to single line in [PR #6](https://github.com/tautschnig/cbmc/pull/6)

thk123 commented on src/cpp/cpp_instantiate_template.cpp:
```diff
@@ -11,6 +11,12 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
 #include "cpp_typecheck.h"
 
+//#define DEBUG
```
Nit: don't check this line in ( @peterschrammel )

thk123 commented on src/cpp/cpp_typecheck_compound_type.cpp:
```diff
@@ -11,6 +11,12 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
 #include "cpp_typecheck.h"
 
+//#define DEBUG
```
Nit: don't check this line in ( @peterschrammel )

peterschrammel replied:
Agreed (also with the other files where this occurs).

andreast271 replied:
These are already fixed in the latest commit

thk123 commented on src/cpp/cpp_typecheck_compound_type.cpp:
```diff
@@ -883,12 +889,25 @@ void cpp_typecheckt::typecheck_friend_declaration(
+	      << sub_it->value().pretty() << std::endl;
+    std::cout << "  scope: " << cpp_scopes.current_scope().prefix << std::endl;
+#endif
+    // In which scope are we going to typecheck this?
```
Add a comment explaining why the following code is ifdef'd out but not deleted

thk123 commented on src/cpp/cpp_typecheck_method_bodies.cpp:
```diff
@@ -9,19 +9,30 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 /// \file
 /// C++ Language Type Checking
 
+//#define DEBUG
```
Nit: don't check this in ( @peterschrammel )

thk123 commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -11,6 +11,12 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 
 #include "cpp_typecheck_resolve.h"
 
+//#define DEBUG
```
Nit: don't check in ( @peterschrammel )

thk123 commented on src/cpp/cpp_typecheck.cpp:
```diff
@@ -323,3 +328,13 @@ void cpp_typecheckt::clean_up()
+{
+  if(expr.id()==ID_cpp_name)
+    return true;
+  forall_operands(it, expr)
```
Nit: can you use a range based for loop here? Also since for loop extends over multiple lines please wrap in braces

andreast271 replied:
Done in [PR #6](https://github.com/tautschnig/cbmc/pull/6)

thk123 commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -11,7 +11,7 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 #include "cpp_typecheck_resolve.h"
 
-//#define DEBUG
+// #define DEBUG
```
Nit: remove this line ( @peterschrammel )

thk123 commented on src/cpp/cpp_typecheck_resolve.cpp:
```diff
@@ -486,51 +486,44 @@ void cpp_typecheck_resolvet::disambiguate_functions(
-          identifiers.begin();
-          it2!=identifiers.end();
-          ) // no it2++
+      for(resolve_identifierst::const_iterator it2=it1+1;
```
Nit: giving these iterators meaningful names might make this code a bit easier to follow

thk123 commented:
Did review commit by commit, so might still be worth checking the comments that are now out-dated, but they may be fixed in later commits

tautschnig commented:
@andreast271 I've addressed the remaining concerns voiced by @thk123. I believe this would not warrant a final (?) review by @peterschrammel and should then be merged. There is of course a lot of work left to be done in this front-end...

thk123 commented on regression/cpp/gcc_vector1/main.cpp:
```diff
@@ -0,0 +1,22 @@
+
+#else
+
+void setzero()
```
How come we're only testing on GNUC?

tautschnig replied:
I have no idea whether Visual Studio supports (the same) vector types?

andreast271 replied:
MSVC doesn't support the vector_size attribute. It does have [128-bit data types](https://docs.microsoft.com/en-us/cpp/cpp/m128). However, these are defined in header files, and the cbmc C++ front end cannot parse MSVC C++ header files.

tautschnig replied:
Is the syntax to use them the same as with GCC/Clang? Would you actually be able to contribute a `KNOWNBUG` regression test that uses them?

andreast271 replied:
The existing testcase uses the typedef for __m128 from the xmmintrin.h header files that ships with gcc. I can do the same for MSVC and add that to the existing test case in a separate PR (after this one is merged). There is an interesting problem, though: CBMC with `PARSER.mode==configt::ansi_ct::flavourt::VISUAL_STUDIO` accepts the syntax of the test case, but the MSVC compiler does not: it complains that ```error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax```
Should cbmc in this case follow MSVC and report a parser error?

tautschnig replied:
It would be desirable if the CBMC front-end mimicked exactly what MSVC does, but that's surely not of the highest priority.

thk123 commented on src/cpp/parse.cpp:
```diff
@@ -4401,6 +4401,16 @@ bool Parser::rClassSpec(typet &spec)
     if(!optAlignas(spec))
       return false;
+
+    if(lex.LookAhead(0)==TOK_GCC_ATTRIBUTE)
```
I don't know what this means exactly, but it would probably make sense to have some tests off GCC attributes after struct/class declarations.

tautschnig replied:
I'll try to dig up with those tests were: we do have them already, otherwise the change wouldn't have happened :-) - but then it might be rather implicit in that it's drawn in from system headers.

thk123 commented on regression/cbmc/cpp1/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+^warning: ignoring
```
Link to the issue would be great

tautschnig replied:
Done.

thk123 commented on regression/cbmc/cpp2/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+--
+^warning: ignoring
+^equality without matching types
```
Nit: Link to the issue

tautschnig replied:
Done.

thk123 commented on src/cpp/cpp_convert_type.cpp:
```diff
@@ -130,6 +130,10 @@ void cpp_convert_typet::read_rec(const typet &type)
     extern_cnt++;
+  else if(type.id()==ID_noreturn)
+  {
+    noreturn_cnt++;
```
Nit: prefer prefix increment for consistency 
```c++
++noreturn_cnt;
```

tautschnig replied:
But ... that wouldn't be _consistent_ here?

thk123 commented on src/cpp/cpp_convert_type.cpp:
```diff
@@ -28,7 +28,7 @@ class cpp_convert_typet
            long_cnt, const_cnt, restrict_cnt, constexpr_cnt, volatile_cnt,
            double_cnt, float_cnt, complex_cnt, cpp_bool_cnt, proper_bool_cnt,
-           extern_cnt, wchar_t_cnt, char16_t_cnt, char32_t_cnt,
+           extern_cnt, noreturn_cnt, wchar_t_cnt, char16_t_cnt, char32_t_cnt,
```
As an aside - variable names should not be abbreviated, though I don't know whether it is better to un-abbreviate this new one whilst leaving the old as they are or doing a separate PR to fix them all.

tautschnig commented:
@peterschrammel, @kroening  ping?

tautschnig commented:
@kroening would you mind taking a look?

tautschnig commented:
@kroening ping?

kroening commented:
- [x] Finish my changes
- [ ] Push my commits to GitHub
- [ ] Open a pull request

tautschnig commented:
@kroening This is finally ready for review - it used to be a monster, but is now down to 2 commits with just +90/−26 modified lines.
